File: 2-Graph Search, Shortest Path, and Data Structures/3-Graph Search/BFS : DFS/Python/bfs_and_dfs_apps.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

__author__ = 'Ziang Lu'

from directed_graph import DirectedGraph
from undirected_graph import UndirectedGraph


def main():
    """
    Main driver.
    :return: None
    """
    print('Testing undirected graph...')
    _test_undirected_graph(filename='undirected_graph_info.txt')
    print()

    print('Testing directed graph...')
    _test_directed_graph(filename='directed_graph_info.txt')


def _test_undirected_graph(filename):
    """
    Test driver for undirected graph with BFS.
    :param filename: str
    :return: None
    """
    # Construct the graph
    graph = _construct_graph(filename=filename, undirected=True)

    # BFS

    # Find all the findable vertices starting from vertex #1 using BFS
    print('Findable vertices from vertex #1 using BFS: %s' %
          graph.bfs(src_vtx_id=1))  # [1, 2, 3, 4, 5, 6]
    graph.clear_explored()

    # Find the length of the shortest path from vertex #1 to vertex #6
    print('Length of the shortest path from vertex #1 to vertex #6: %d' %
          graph.shortest_path(src_vtx_id=1, dest_vtx_id=6))  # 3
    graph.clear_explored()

    # Find the number of connected components of the undirected graph using BFS
    # (Undirected connectivity)
    print('Number of connected components of the graph using BFS (Undirected '
          'connectivity): %d' % graph.num_of_connected_components_with_bfs())  # 1
    graph.clear_explored()

    # DFS

    # Find all the findable vertices starting from vertex #1 using DFS
    print('Findable vertices from vertex #1 using DFS: %s' %
          graph.bfs(src_vtx_id=1))  # [1, 2, 3, 4, 5, 6]
    graph.clear_explored()

    # Find the number of connected components of the undirected graph using DFS
    # (Undirected connectivity)
    print('Number of connected components of the graph using DFS (Undirected '
          'connectivity): %d' % graph.num_of_connected_components_with_dfs())  # 1
    graph.clear_explored()


def _test_directed_graph(filename):
    """
    Test driver for directed graph with BFS.
    :param filename: str
    :return: None
    """
    # Construct the graph
    graph = _construct_graph(filename=filename, undirected=False)

    # BFS

    # Find all the findable vertices from vertex #1 using BFS
    print('Findable vertices from vertex #1 using BFS: %s' %
          graph.bfs(src_vtx_id=1))  # [1, 4, 7]
    graph.clear_explored()

    # Find the length of the shortest path from vertex #1 to vertex #7
    print('Length of the shortest path from vertex #1 to vertex #7: %d' %
          graph.shortest_path(src_vtx_id=1, dest_vtx_id=7))  # 2
    graph.clear_explored()

    # DFS

    # Find all the findable vertices from vertex #1 using DFS
    print('Findable vertices from vertex #1 using DFS: %s' %
          graph.bfs(src_vtx_id=1))  # [1, 4, 7]
    graph.clear_explored()

    # Find the number of SCCs of the directed graph using DFS
    # (Directed connectivity)
    print('Number of SCCs of the graph using DFS (Directed connectivity): %d' %
          graph.num_of_connected_components_with_dfs())  # 3
    graph.clear_explored()

    # Find the topological ordering of the vertices of the directed graph using
    # DFS
    print('Topological ordering of the vertices of the directed graph using '
          'DFS: %s' % graph.topological_sort())  # [2, 8, 5, 6, 9, 3, 1, 4, 7]
    # The result might be wrong when there is no topological ordering of the
    # directed graph.

    # Find the topological ordering of the vertices of the directed graph using
    # straightforward algorithm
    print('Topological ordering of the vertices of the directed graph using '
          'straightforward algorithm: %s' %
          graph.topological_sort_straightforward())  # [0, 0, 0, 0, 0, 0, 0, 0, 0]
    # The result is because that the graph doesn't have a topological ordering.


def _construct_graph(filename, undirected):
    """
    Private helper method to construct a graph from the given graph file.
    :param filename: str
    :param undirected: bool
    :return: UndirectedGraph or DirectedGraph
    """
    with open(filename, 'rt') as f:
        # Construct the graph
        if undirected:
            graph = UndirectedGraph()
        else:
            graph = DirectedGraph()
        # Add the vertices
        n_vtx = int(f.readline())
        for vtx_id in range(1, n_vtx + 1):
            graph.add_vtx(new_vtx_id=vtx_id)
        # Add the edges
        for line in f.readlines():
            ends = line.split(' ')
            graph.add_edge(int(ends[0]), int(ends[1]))
        return graph


if __name__ == '__main__':
    main()

File: 2-Graph Search, Shortest Path, and Data Structures/3-Graph Search (BFS & DFS)/Java/BFSAndDFSApps.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Arrays;
import java.util.Scanner;

import directed_graph.DirectedGraph;
import graph.GraphInterface;
import undirected_graph.UndirectedGraph;

public class BFSAndDFSApps {

    /**
     * Main driver.
     * @param args arguments from command line
     */
    public static void main(String[] args) {
        System.out.println("Testing undirected graph...");
        testUndirectedGraph("undirected_graph_info.txt");
        System.out.println();

        System.out.println("Testing directed graph...");
        testDirectedGraph("directed_graph_info.txt");
    }

    /**
     * Test driver for BFS with undirected graph.
     * @param filename undirected graph filename
     */
    private static void testUndirectedGraph(String filename) {
        GraphInterface graph = constructGraph(filename, true);

        // BFS

        // Find all the findable vertices starting from vertex #1 using BFS
        System.out.println("Findable vertices from vertex #1 using BFS: " + graph.bfs(1)); // [1, 2, 3, 4, 5, 6]
        graph.clearExplored();

        // Find the length of the shortest path from vertex #1 to vertex #6
        System.out.println("Length of the shortest path from vertex #1 to vertex #6: " + graph.shortestPath(1, 6)); // 3
        graph.clearExplored();

        // Find the number of connected components of the undirected graph using BFS (Undirected connectivity)
        System.out.println("Number of connected components of the graph using BFS (Undirected connectivity): "
                + graph.numOfConnectedComponentsWithBFS()); // 1
        graph.clearExplored();

        // DFS

        // Find all the findable vertices starting from vertex #1 using DFS
        System.out.println("Findable vertices from vertex #1 using DFS: " + graph.dfs(1)); // [1, 2, 3, 4, 5, 6]
        graph.clearExplored();

        // Find the number of connected components of the undirected graph using DFS (Undirected connectivity)
        System.out.println("Number of connected components of the graph using DFS (Undirected connectivity): "
                + graph.numOfConnectedComponentsWithDFS()); // 1
        graph.clearExplored();
    }

    /**
     * Test driver for BFS with directed graph.
     * @param filename directed graph filename
     */
    private static void testDirectedGraph(String filename) {
        DirectedGraph graph = (DirectedGraph) constructGraph(filename, false);

        // Find all the findable vertices starting from vertex #1 using BFS
        System.out.println("Findable vertices from vertex #1 using BFS: " + graph.bfs(1)); // [1, 4, 7]
        graph.clearExplored();

        // Find the length of the shortest path from vertex #1 to vertex #7
        System.out.println("Length of the shortest path from vertex #1 to vertex #7: " + graph.shortestPath(1, 7)); // 2
        graph.clearExplored();

        // DFS

        // Find all the findable vertices starting from vertex #1 using DFS
        System.out.println("Findable vertices from vertex #1 using DFS: " + graph.dfs(1)); // [1, 4, 7]
        graph.clearExplored();

        // Find the number of SCCs of the directed graph using DFS (Directed connectivity)
        System.out.println("Number of SCCs of the graph using DFS (Directed connectivity): "
                + graph.numOfConnectedComponentsWithDFS()); // 3
        graph.clearExplored();

        // Find the topological ordering of the vertices of the directed graph using DFS
        System.out.println("Topological ordering of the vertices of the directed graph using DFS: "
                + Arrays.toString(graph.topologicalSort())); // [2, 8, 5, 6, 9, 3, 1, 4, 7]
        graph.clearExplored();
        // The result might be wrong when there is no topological ordering of the directed graph.

        // Find the topological ordering of the vertices of the directed graph using straightforward algorithm
        System.out
                .println("Topological ordering of the vertices of the directed graph using straightforward algorithm: "
                        + Arrays.toString(graph.topologicalSortStraightforward())); // [0, 0, 0, 0, 0, 0, 0, 0, 0]
        graph.clearExplored();
        // The result is because that the graph doesn't have a topological ordering.
    }

    /**
     * Private helper method to construct a graph from the given file.
     * @param filename given filename
     * @param undirected whether the graph to construct is undirected
     * @return constructed undirected graph
     */
    private static GraphInterface constructGraph(String filename, boolean undirected) {
        GraphInterface graph = null;

        Scanner scanner = null;
        try {
            scanner = new Scanner(new File(filename), "latin1");

            if (undirected) {
                graph = new UndirectedGraph();
            } else {
                graph = new DirectedGraph();
            }
            // Add the vertices
            int nVtx = Integer.parseInt(scanner.nextLine());
            for (int vtxID = 1; vtxID <= nVtx; ++vtxID) {
                graph.addVtx(vtxID);
            }
            // Add the edges
            while (scanner.hasNextLine()) {
                String[] ends = scanner.nextLine().split(" ");
                graph.addEdge(Integer.parseInt(ends[0]), Integer.parseInt(ends[1]));
            }
        } catch (FileNotFoundException ex) {
            System.out.println("Cannot found the file.");
        } finally {
            if (scanner != null) {
                scanner.close();
            }
        }

        return graph;
    }

}
